File: algorithm/프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def is_changable(src, dst):
    cnt = 0
    for i in range(len(src)):
        if src[i] != dst[i]:
            cnt += 1
            
    return cnt == 1

def solution(begin, target, words):
    dist = {word: 999 for word in words}
    dist[begin] = 0
    
    q = deque()
    q.append(begin)
    while q:
        w = q.popleft()
        if w == target:
            return dist[w]
        
        for word in words:
            if is_changable(w, word) and dist[word] > dist[w]:
                dist[word] = dist[w] + 1
                q.append(word)
        
    return 0

File: algorithm/프로그래머스/3/43163. 단어 변환/단어 변환.go
func isChangable(src, dst string) bool {
    cnt := 0
    for i, _ := range src {
        if src[i] != dst[i] {
            cnt += 1
        }
    }
    
    return cnt == 1
}

func solution(begin string, target string, words []string) int {
    dist := make(map[string]int, len(words) + 1)
    for _, word := range words {
        dist[word] = 999
    }
    dist[begin] = 0
    
    q := []string{begin}
    for len(q) > 0 {
        w := q[0]
        q = q[1:]
        
        if w == target {
            return dist[w]
        }
        
        for _, word := range words {
            if isChangable(w, word) && dist[word] > dist[w] {
                q = append(q, word)
                dist[word] = dist[w] + 1
            }
        }
    }
    
    return 0
}